File: backend/src/Fix.Domain/AggregateRoots/Orders/Order.cs
using Fix.Domain.Abstractions;
using Fix.Domain.Common;
using Fix.Domain.AggregateRoots.Counterparties;
using Fix.Domain.AggregateRoots.Mandates;

namespace Fix.Domain.AggregateRoots.Orders;

/// <summary>
/// Boleta de hedge (order): executa e consome um mandato ativo. Futuros e opções consomem lotes;
/// NDF consome nocional em US$. Boletas emitidas sem alçada aguardam aprovação e não consomem saldo.
/// </summary>
public sealed class Order : AggregateRoot, IOrganizationScoped
{
    /// <summary>Prazo (dias úteis, aprox.) para o confirmation chegar antes de ficar em atraso.</summary>
    public const int ConfirmationDeadlineBusinessDays = 2;

    private Order()
    {
    }

    private Order(Guid organizationId, Guid mandateId, Guid requestedBy)
    {
        OrganizationId = organizationId;
        MandateId = mandateId;
        RequestedBy = requestedBy;
        Confirmation = ConfirmationStatus.Pending;
    }

    public Guid OrganizationId { get; private set; }

    public Guid MandateId { get; private set; }

    public OrderType Type { get; private set; }

    public TradeDirection Direction { get; private set; }

    public Guid CounterpartyId { get; private set; }

    public Commodity? Commodity { get; private set; }

    public Tenor Tenor { get; private set; } = null!;

    /// <summary>Lotes (futuros e opções).</summary>
    public decimal? Lots { get; private set; }

    /// <summary>Nocional em US$ (NDF).</summary>
    public decimal? NotionalUsd { get; private set; }

    /// <summary>Preço (futuro), taxa (NDF) ou strike (opção).</summary>
    public decimal Price { get; private set; }

    public string PriceUnit { get; private set; } = null!;

    public OptionKind? OptionKind { get; private set; }

    public decimal? Premium { get; private set; }

    public DateOnly TradeDate { get; private set; }

    public string? Notes { get; private set; }

    public ApprovalStatus Approval { get; private set; }

    public Guid RequestedBy { get; private set; }

    public Guid? DecidedBy { get; private set; }

    public DateTimeOffset? DecidedAt { get; private set; }

    public string? DecisionNote { get; private set; }

    public ConfirmationStatus Confirmation { get; private set; }

    public DateOnly? ConfirmedOn { get; private set; }

    public string? ConfirmationNote { get; private set; }

    /// <summary>Quantidade que consome o mandato (lotes ou US$). Pendentes e rejeitadas não consomem.</summary>
    public decimal ConsumedQuantity => Approval == ApprovalStatus.Approved ? Quantity : 0;

    public decimal Quantity => Type == OrderType.Ndf ? NotionalUsd ?? 0 : Lots ?? 0;

    public static Order Register(
        Mandate mandate,
        Counterparty counterparty,
        OrderTerms terms,
        decimal consumedByOthers,
        Guid requestedBy,
        bool requesterHasAuthority)
    {
        var order = new Order(mandate.OrganizationId, mandate.Id, requestedBy);
        order.Apply(mandate, counterparty, terms, consumedByOthers);
        order.Approval = requesterHasAuthority ? ApprovalStatus.Approved : ApprovalStatus.PendingApproval;
        return order;
    }

    /// <summary>Permitido enquanto pendente de aprovação ou aprovada sem confirmation.</summary>
    public void Update(Mandate mandate, Counterparty counterparty, OrderTerms terms, decimal consumedByOthers)
    {
        if (Approval == ApprovalStatus.Rejected || Confirmation != ConfirmationStatus.Pending)
        {
            throw new DomainException("Boletas rejeitadas ou com confirmation registrado não podem ser editadas.");
        }

        Apply(mandate, counterparty, terms, consumedByOthers);
    }

    public void Approve(Guid decidedBy, DateTimeOffset at, string? note)
    {
        EnsurePendingApproval();
        Approval = ApprovalStatus.Approved;
        Decide(decidedBy, at, DomainGuard.OptionalText(note, 500, "observação"));
    }

    public void Reject(Guid decidedBy, DateTimeOffset at, string reason)
    {
        EnsurePendingApproval();
        Approval = ApprovalStatus.Rejected;
        Decide(decidedBy, at, DomainGuard.OptionalText(reason, 500, "justificativa")
            ?? throw new DomainException("A rejeição exige justificativa."));
    }

    public bool CanBeDeleted => Approval != ApprovalStatus.Approved;

    // ---------- Confirmation (middle office) ----------

    public void Confirm(DateOnly receivedOn)
    {
        EnsureApproved();
        if (receivedOn < TradeDate)
        {
            throw new DomainException("O confirmation não pode ser anterior à data do trade.");
        }

        Confirmation = ConfirmationStatus.Confirmed;
        ConfirmedOn = receivedOn;
        ConfirmationNote = null;
    }

    public void MarkDivergent(string description)
    {
        EnsureApproved();
        Confirmation = ConfirmationStatus.Divergent;
        ConfirmationNote = DomainGuard.OptionalText(description, 500, "divergência")
            ?? throw new DomainException("A divergência exige descrição.");
    }

    public void RefuseConfirmation(string reason)
    {
        EnsureApproved();
        Confirmation = ConfirmationStatus.Refused;
        ConfirmationNote = DomainGuard.OptionalText(reason, 500, "motivo")
            ?? throw new DomainException("A recusa exige motivo.");
    }

    /// <summary>Divergência sanada: confirmation aceito.</summary>
    public void ResolveDivergence(DateOnly on)
    {
        if (Confirmation is not (ConfirmationStatus.Divergent or ConfirmationStatus.Refused))
        {
            throw new DomainException("Não há divergência a sanar nesta boleta.");
        }

        Confirmation = ConfirmationStatus.Confirmed;
        ConfirmedOn ??= on;
        ConfirmationNote = null;
    }

    /// <summary>Pendente além do prazo (dias corridos convertidos de forma aproximada para dias úteis).</summary>
    public bool IsConfirmationOverdue(DateOnly today)
    {
        if (Approval != ApprovalStatus.Approved || Confirmation != ConfirmationStatus.Pending)
        {
            return false;
        }

        var businessDays = 0;
        for (var day = TradeDate.AddDays(1); day <= today; day = day.AddDays(1))
        {
            if (day.DayOfWeek is not (DayOfWeek.Saturday or DayOfWeek.Sunday))
            {
                businessDays++;
            }
        }

        return businessDays > ConfirmationDeadlineBusinessDays;
    }

    // ---------- Regras ----------

    private void Apply(Mandate mandate, Counterparty counterparty, OrderTerms terms, decimal consumedByOthers)
    {
        if (mandate.Id != MandateId)
        {
            throw new DomainException("A boleta não pode trocar de mandato.");
        }

        if (!mandate.AcceptsOrders)
        {
            throw new DomainException("O mandato não está ativo — pendente, rejeitado ou encerrado não aceita boleta.");
        }

        if (!counterparty.IsHomologated)
        {
            throw new DomainException($"A contraparte {counterparty.Name} não está homologada (§5.5).");
        }

        var expected = mandate.Type switch
        {
            MandateType.Pricing => terms.Type is OrderType.Futures or OrderType.Option,
            MandateType.Currency => terms.Type is OrderType.Ndf,
            _ => false,
        };
        if (!expected)
        {
            throw new DomainException(mandate.Type is MandateType.Currency
                ? "Mandatos de moeda aceitam apenas boletas de NDF."
                : mandate.Type is MandateType.Pricing
                    ? "Mandatos de precificação aceitam apenas futuros e opções."
                    : "Mandatos comerciais e logísticos são executados por contratos, não por boletas de hedge.");
        }

        if (terms.Type == OrderType.Ndf)
        {
            if (terms.NotionalUsd is not > 0)
            {
                throw new DomainException("Informe o nocional em US$ do NDF.");
            }
        }
        else if (terms.Lots is not > 0)
        {
            throw new DomainException("Informe a quantidade de lotes.");
        }

        if (terms.Price <= 0)
        {
            throw new DomainException(terms.Type == OrderType.Option ? "Informe o strike da opção." : "Informe o preço.");
        }

        if (terms.Type == OrderType.Option && (terms.OptionKind is null || terms.Premium is not >= 0))
        {
            throw new DomainException("Opções exigem o tipo (put/call) e o prêmio.");
        }

        var quantity = terms.Type == OrderType.Ndf ? terms.NotionalUsd!.Value : terms.Lots!.Value;
        var consumes = terms.Type == OrderType.Ndf ? MeasurementUnit.Usd : MeasurementUnit.Lots;
        if (mandate.Quantity is { } authorized && mandate.QuantityUnit == consumes && consumedByOthers + quantity > authorized)
        {
            throw new DomainException(
                $"A boleta excede o saldo do mandato (autorizado {authorized:N0}, consumido {consumedByOthers:N0}, boleta {quantity:N0}).");
        }

        Type = terms.Type;
        Direction = terms.Direction;
        CounterpartyId = counterparty.Id;
        Commodity = terms.Type == OrderType.Ndf ? null : mandate.Commodity;
        Tenor = terms.Tenor;
        Lots = terms.Type == OrderType.Ndf ? null : terms.Lots;
        NotionalUsd = terms.Type == OrderType.Ndf ? terms.NotionalUsd : null;
        Price = terms.Price;
        PriceUnit = DomainGuard.OptionalText(terms.PriceUnit, 20, "unidade de preço")
            ?? (terms.Type == OrderType.Ndf ? "R$/US$" : "c/lb");
        OptionKind = terms.Type == OrderType.Option ? terms.OptionKind : null;
        Premium = terms.Type == OrderType.Option ? terms.Premium : null;
        TradeDate = terms.TradeDate;
        Notes = DomainGuard.Description(terms.Notes);
    }

    private void EnsurePendingApproval()
    {
        if (Approval != ApprovalStatus.PendingApproval)
        {
            throw new DomainException("Esta boleta não está pendente de aprovação.");
        }
    }

    private void EnsureApproved()
    {
        if (Approval != ApprovalStatus.Approved)
        {
            throw new DomainException("O confirmation só se aplica a boletas aprovadas.");
        }
    }

    private void Decide(Guid decidedBy, DateTimeOffset at, string? note)
    {
        DecidedBy = decidedBy;
        DecidedAt = at;
        DecisionNote = note;
    }
}

/// <summary>Termos editáveis da boleta.</summary>
public sealed record OrderTerms(
    OrderType Type,
    TradeDirection Direction,
    Tenor Tenor,
    decimal? Lots,
    decimal? NotionalUsd,
    decimal Price,
    string? PriceUnit,
    OptionKind? OptionKind,
    decimal? Premium,
    DateOnly TradeDate,
    string? Notes);

File: frontend/web/src/domain/order.ts
import type { Commodity } from './common';
import type { MandateType } from './mandate';

export const ORDER_TYPES = ['Futures', 'Option', 'Ndf'] as const;
export type OrderType = (typeof ORDER_TYPES)[number];

export const TRADE_DIRECTIONS = ['Buy', 'Sell'] as const;
export type TradeDirection = (typeof TRADE_DIRECTIONS)[number];

export const OPTION_KINDS = ['Call', 'Put'] as const;
export type OptionKind = (typeof OPTION_KINDS)[number];

export const APPROVAL_STATUSES = ['PendingApproval', 'Approved', 'Rejected'] as const;
export type ApprovalStatus = (typeof APPROVAL_STATUSES)[number];

export const CONFIRMATION_STATUSES = ['Pending', 'Confirmed', 'Divergent', 'Refused'] as const;
export type ConfirmationStatus = (typeof CONFIRMATION_STATUSES)[number];

/** Boletas executam mandatos de precificação (futuros/opções) e de moeda (NDF). */
export const orderTypesFor = (type: MandateType): OrderType[] =>
  type === 'Pricing' ? ['Futures', 'Option'] : type === 'Currency' ? ['Ndf'] : [];

export interface OrderTerms {
  type: OrderType;
  direction: TradeDirection;
  /** Vencimento: N26, fev/27. */
  tenor: string;
  /** Futuros e opções. */
  lots: number | null;
  /** NDF. */
  notionalUsd: number | null;
  /** Preço (futuro), taxa (NDF) ou strike (opção). */
  price: number;
  priceUnit: string | null;
  optionKind: OptionKind | null;
  premium: number | null;
  tradeDate: string;
  notes: string | null;
  /** Venda de opção coberta; a descoberta é vedada (FORA). */
  coveredSale: boolean;
}

/** Termos enviados ao registrar/editar: commodity só vale sem mandato; justificativa é obrigatória em desvio. */
export interface OrderTermsInput extends OrderTerms {
  commodity: Commodity | null;
  justification: string | null;
}

/** Enquadramento calculado da boleta (FIX2: desvio não bloqueia — expõe). */
export interface OrderCompliance {
  status: 'Within' | 'Outside';
  reason: string;
}

export interface Order {
  id: string;
  /** HX-0001 */
  code: string;
  /** Nulo = boleta sem mandato (desvio sinalizado). */
  mandateId: string | null;
  /** MD-01 */
  mandateCode: string | null;
  mandateTitle: string;
  counterpartyId: string;
  counterpartyName: string;
  /** Ausente em NDF / em mandatos que não são de precificação. */
  commodity: Commodity | null;
  terms: OrderTerms;
  approval: ApprovalStatus;
  requestedBy: string;
  decidedBy: string | null;
  decidedAt: string | null;
  decisionNote: string | null;
  confirmation: ConfirmationStatus;
  confirmedOn: string | null;
  confirmationNote: string | null;
  /** Pendente há mais de 2 dias úteis. */
  confirmationOverdue: boolean;
  /** Middle office que registrou a confirmação (nunca quem executou). */
  confirmationBy: string | null;
  compliance: OrderCompliance;
  /** Mandato vinculado depois da execução: carimbo permanente. */
  linkedAfterExecution: boolean;
  exceedsMandate: boolean;
  deviationNote: string | null;
}

/** Carimbos de desvio visíveis na boleta (FIX2 · I-01). */
export function orderStamps(order: Order): { label: string; tone: 'danger' | 'warning' | 'info' }[] {
  const stamps: { label: string; tone: 'danger' | 'warning' | 'info' }[] = [];
  if (!order.mandateId) stamps.push({ label: 'sem mandato', tone: 'danger' });
  if (order.linkedAfterExecution) stamps.push({ label: 'a posteriori', tone: 'warning' });
  if (order.exceedsMandate) stamps.push({ label: 'estourou o mandato', tone: 'danger' });
  if (order.compliance.status === 'Outside' && stamps.length === 0) stamps.push({ label: 'FORA', tone: 'danger' });
  return stamps;
}

/** Confirmação em aberto: boleta aprovada sem confirmação conferida. */
export const hasOpenConfirmation = (order: Order) => order.approval === 'Approved' && order.confirmation !== 'Confirmed';

/** Termos em edição no formulário (preço ainda pode estar vazio). */
export type OrderTermsDraft = Omit<OrderTerms, 'price'> & { price: number | null };
